compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/scene/ComposeScene.kt

```diff
@@ -0,0 +1,267 @@
+import androidx.compose.ui.input.key.KeyEvent
+import androidx.compose.ui.input.key.KeyInputModifierNode
+import androidx.compose.ui.input.pointer.InteropViewCatchPointerModifier
+import androidx.compose.ui.input.pointer.PointerButton
+import androidx.compose.ui.input.pointer.PointerButtons
+import androidx.compose.ui.input.pointer.PointerEventType
+import androidx.compose.ui.input.pointer.PointerKeyboardModifiers
+import androidx.compose.ui.input.pointer.PointerType
+import androidx.compose.ui.node.LayoutNode
+import androidx.compose.ui.platform.LocalLayoutDirection
+import androidx.compose.ui.platform.setContent
+import androidx.compose.ui.semantics.SemanticsNode
+import androidx.compose.ui.semantics.SemanticsOwner
+import androidx.compose.ui.unit.Constraints
+import androidx.compose.ui.unit.Density
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.IntRect
+import androidx.compose.ui.unit.IntSize
+import androidx.compose.ui.unit.LayoutDirection
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.window.Dialog
+import androidx.compose.ui.window.Popup
+import org.jetbrains.skiko.currentNanoTime
+
+/**
+ * Represents a static [CompositionLocal] key for a [ComposeScene] in Jetpack Compose.
+ *
+ * @see ComposeScene
+ */
+internal val LocalComposeScene = staticCompositionLocalOf<ComposeScene?> { null }
+
+/**
+ * The local [ComposeScene] is typically not-null. This extension can be used in these cases.
+ */
+@Composable
+internal fun CompositionLocal<ComposeScene?>.requireCurrent(): ComposeScene {
+    return current ?: error("CompositionLocal LocalComposeScene not provided")
+}
+
+/**
+ * A virtual container that encapsulates Compose UI content. UI content can be constructed via
+ * [setContent] method and with any Composable that manipulates [LayoutNode] tree.
+ *
+ * After [ComposeScene] will no longer needed, you should call [close] method, so all resources
+ * and subscriptions will be properly closed. Otherwise, there can be a memory leak.
+ */
+@InternalComposeUiApi
+interface ComposeScene {
+    /**
+     * Density of the content which will be used to convert [Dp] units.
+     */
+    var density: Density
+
+    /**
+     * The layout direction of the content, provided to the composition via [LocalLayoutDirection].
+     */
+    var layoutDirection: LayoutDirection
+
+    /**
+     * Represents the scene size in scaled pixels ([Dp]).
+     * This is used to impose constraints on the content. If the size is undefined, it can be
+     * set to `null`. In such a case, the content will be laid out without any restrictions and
+     * the window size will be utilized to bounds verification.
+     */
+    var size: IntSize?
+
+    /**
+     * The object that used to share "context" between multiple scenes on the screen.
+     * Also, it provides a way for platform interaction that required within a scene.
+     *
+     * @see ComposeSceneContext
+     */
+    val composeSceneContext: ComposeSceneContext
```

Comment:
```suggestion
    // TODO remove and provide LocalPlatformContext for all usages
    val composeSceneContext: ComposeSceneContext
```

As discussed, better to not provide an API in ComposeScene that should be used by ComposeScene's content. ComposeScene should has API only for outside client.